src/lib/cache/helper.go

```diff
@@ -0,0 +1,88 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/goharbor/harbor/src/lib/log"
+)
+
+var (
+	onceMu = keyMutex{m: &sync.Map{}}
+)
+
+// Once retrieve the cached key value, build and cache the value when the cached key value not found
+func Once(c Cache, key string, value interface{}, builder func() (interface{}, error), expiration ...time.Duration) error {
+	err := c.Fetch(key, value)
+	// value found from the cache
+	if err == nil {
+		return nil
+	}
+	// internal error
+	if !errors.Is(err, ErrNotFound) {
+		return err
+	}
+
+	// lock the key in cache and try to build the value for the key
+	onceKey := fmt.Sprintf("%p:%s", c, key)
+	onceMu.Lock(onceKey)
+
+	defer onceMu.Unlock(onceKey)
+
+	// fetch again to avoid to build the value multi-times
+	err = c.Fetch(key, value)
+	if err == nil {
+		return nil
+	}
+	// internal error
+	if !errors.Is(err, ErrNotFound) {
+		return err
+	}
+
+	val, err := builder()
+	if err != nil {
+		return err
+	}
+
+	if err := c.Save(key, val, expiration...); err != nil {
+		log.Warningf("failed to save value to cache, error: %v", err)
+
+		// save the val to cache failed, copy it to the value directly
+		return simpleCopy(value, val)
+	}
+
+	return c.Fetch(key, value) // after the building, fetch value again
+}
+
+// OnceContext executes Once function when Cache found in ctx, otherwise builds and assigns the result to value
+func OnceContext(ctx context.Context, key string, value interface{}, builder func() (interface{}, error), expiration ...time.Duration) error {
```

Comment:
Not quite sure why this is needed.

Reply:
This is preparing for the DAOs. For example, adding a cache layer to the `Get` method of the DAO for the project may like this.


``` golang
func(d *dao) Get(ctx context.Context, id int64) (*models.Project, error) {
    key := fmt.Sprintf("project:%d", id)
    var project models.Project
    err := cache.OnceContext(ctx, key, &project, func() (interface{}, error) error {
        o, err := orm.FromContext(ctx)
        if err != nil {
            return nil, err
        }
	p := &models.Project{ProjectID: id, Deleted: false}
	if err = o.Read(p, "project_id", "deleted"); err != nil {
		return nil, orm.WrapNotFoundError(err, "project %d not found", id)
	}

        return p, nil
    })

   if err != nil {
       return nil, err
   }

   return &project, nil
}
```